File: macos/Sources/VoiceKeyCore/TranscriptCleaner.swift
import Foundation

/// Post-processing for raw speech-to-text output: strips disfluencies that
/// Whisper faithfully transcribes, and recognises whole-transcript non-speech
/// annotations (`[wind howling]`, `(music)`) that must never be pasted.
public enum TranscriptCleaner {

    /// Removes filler words and collapses stutter repetitions. Pure.
    public static func clean(_ text: String) -> String {
        normalisingWhitespace(
            collapsingImmediateRepetitions(
                removingFillers(
                    capitalisingAfterSentenceInitialFillers(text))))
    }

    /// True when the entire transcript is a non-speech annotation like
    /// `[wind howling]` or `(music)`.
    public static func isNonSpeechAnnotation(_ text: String) -> Bool {
        text.trimmingCharacters(in: .whitespacesAndNewlines)
            .wholeMatch(of: #/\[[^\[\]]+\]|\([^()]+\)/#) != nil
    }

    // MARK: - Stages

    /// A capitalised filler opening a sentence hands its capitalisation to the
    /// word that follows, so the filler can then be deleted like any other.
    private static func capitalisingAfterSentenceInitialFillers(_ text: String) -> String {
        text.replacing(sentenceInitialFillerRegex) { match in
            String(match.output[1].substring ?? "") + String(match.output[2].substring ?? "").uppercased()
        }
    }

    private static func removingFillers(_ text: String) -> String {
        text.replacing(fillerRegex, with: "")
    }

    private static func collapsingImmediateRepetitions(_ text: String) -> String {
        let collapsed = repeatedWordRegex.stringByReplacingMatches(
            in: text,
            range: NSRange(location: 0, length: (text as NSString).length),
            withTemplate: "$1"
        )
        return collapsed == text ? text : collapsingImmediateRepetitions(collapsed)
    }

    private static func normalisingWhitespace(_ text: String) -> String {
        text.replacing(#/[ \t]+/#, with: " ")
            .replacing(#/[ \t]*\n[ \t]*/#, with: "\n")
            .trimmingCharacters(in: .whitespacesAndNewlines)
    }

    // MARK: - Regexes

    // Longer spellings precede their prefixes only where `\b` cannot decide;
    // the word boundary makes the order irrelevant for the rest. The alternation
    // is shared by two regexes, so they are built from strings (literals cannot
    // interpolate); the patterns are constant, so the force-try cannot fail once
    // any test has run.
    private static let fillerAlternatives = "um|uhm|umm|uh|erm|er|mhm"

    private static let sentenceInitialFillerRegex = try! Regex(
        #"(\A|[.!?][ \t]+|\n[ \t]*)(?i:\#(fillerAlternatives))\b[ \t]*(?:,|\.\.\.|…)?[ \t]*([a-z])"#
    )

    private static let fillerRegex = try! Regex(
        #"\b(?:\#(fillerAlternatives))\b[ \t]*(?:,|\.\.\.|…)?"#
    ).ignoresCase()

    // NSRegularExpression, not Swift Regex: the Swift engine matches
    // backreferences case-sensitively even under (?i), so "The the" would
    // survive (verified on Swift 6.2.3).
    private static let repeatedWordRegex = try! NSRegularExpression(
        pattern: #"\b([\p{L}\p{N}']+)[ \t]+\1\b"#,
        options: [.caseInsensitive]
    )
}

File: windows/VoiceKey.Core/TranscriptCleaner.cs
using System.Text.RegularExpressions;

namespace VoiceKey.Core;

/// <summary>
/// Post-processing for raw speech-to-text output: strips disfluencies that
/// Whisper faithfully transcribes, and recognises whole-transcript non-speech
/// annotations (<c>[wind howling]</c>, <c>(music)</c>) that must never be pasted.
/// </summary>
public static class TranscriptCleaner
{
    /// <summary>Removes filler words and collapses stutter repetitions. Pure.</summary>
    public static string Clean(string text) =>
        NormalisingWhitespace(
            CollapsingImmediateRepetitions(
                RemovingFillers(
                    CapitalisingAfterSentenceInitialFillers(text))));

    /// <summary>
    /// True when the entire transcript is a non-speech annotation like
    /// <c>[wind howling]</c> or <c>(music)</c>.
    /// </summary>
    public static bool IsNonSpeechAnnotation(string text) =>
        AnnotationRegex.IsMatch(text.Trim());

    // MARK: - Stages

    /// <summary>
    /// A capitalised filler opening a sentence hands its capitalisation to the
    /// word that follows, so the filler can then be deleted like any other.
    /// </summary>
    private static string CapitalisingAfterSentenceInitialFillers(string text) =>
        SentenceInitialFillerRegex.Replace(text,
            match => match.Groups[1].Value + match.Groups[2].Value.ToUpperInvariant());

    private static string RemovingFillers(string text) =>
        FillerRegex.Replace(text, "");

    private static string CollapsingImmediateRepetitions(string text)
    {
        var collapsed = RepeatedWordRegex.Replace(text, "$1");
        return collapsed == text ? text : CollapsingImmediateRepetitions(collapsed);
    }

    private static string NormalisingWhitespace(string text)
    {
        var spaced = HorizontalWhitespaceRegex.Replace(text, " ");
        return PaddedNewlineRegex.Replace(spaced, "\n").Trim();
    }

    // MARK: - Regexes

    // Longer spellings precede their prefixes only where `\b` cannot decide;
    // the word boundary makes the order irrelevant for the rest. The alternation
    // is shared by two expressions.
    private const string FillerAlternatives = "um|uhm|umm|uh|erm|er|mhm";

    private static readonly Regex SentenceInitialFillerRegex = new(
        $@"(\A|[.!?][ \t]+|\n[ \t]*)(?i:{FillerAlternatives})\b[ \t]*(?:,|\.\.\.|…)?[ \t]*([a-z])",
        RegexOptions.Compiled);

    private static readonly Regex FillerRegex = new(
        $@"\b(?:{FillerAlternatives})\b[ \t]*(?:,|\.\.\.|…)?",
        RegexOptions.Compiled | RegexOptions.IgnoreCase);

    private static readonly Regex RepeatedWordRegex = new(
        @"\b([\p{L}\p{N}']+)[ \t]+\1\b",
        RegexOptions.Compiled | RegexOptions.IgnoreCase);

    private static readonly Regex HorizontalWhitespaceRegex = new(
        @"[ \t]+", RegexOptions.Compiled);

    private static readonly Regex PaddedNewlineRegex = new(
        @"[ \t]*\n[ \t]*", RegexOptions.Compiled);

    private static readonly Regex AnnotationRegex = new(
        @"\A(?:\[[^\[\]]+\]|\([^()]+\))\z", RegexOptions.Compiled);
}
